src/lsm/compaction.zig

```diff
@@ -1813,9 +1813,9 @@ pub fn CompactionType(
 
         fn blip_write_next_tick(next_tick: *Grid.NextTick) void {
             const write: *?Beat.Write = @ptrCast(
-                @fieldParentPtr(Beat.Write, "next_tick", next_tick),
+                @as(*Beat.Write, @fieldParentPtr("next_tick", next_tick)),
```

Comment:
:thinking: @cb22 is the original code correct? It casts `*Write` to `*?Write` for the purpose of `fieldParentPtr`, but I think that cast might be invalid? In general, sizeOf(Write) might be unequal to `sizeOf(?Write)`

Reply:
@matklad ha, this was on my mind too recently with Rafael's post in #zig.

It took me a second to figure out WTF was going on here! If we try out a simple test, it panics as you'd expect, because with the current Optional implementation ([not defined!](https://github.com/ziglang/zig/issues/104#issuecomment-824429303)) casting a `*Example` to a `*?Example` and trying to use it results in:

```zig
var example = Example{ .foo = 32 };
var maybe_example: ?Example = example;
var example_pointer: *Example = &example;
var example_maybe_bad_cast: *?Example = @ptrCast(example_pointer);

std.log.info("example={any} maybe_example={any} example_pointer_deref={any} example_maybe_bad_cast_deref={any}", .{
    std.mem.asBytes(&example).*,
    std.mem.asBytes(&maybe_example).*,
    std.mem.asBytes(example_pointer).*,
    std.mem.asBytes(example_maybe_bad_cast).*,
});

var val = example_maybe_bad_cast.*.?;
```

```
info: example={ 32, 0, 0, 0 } maybe_example={ 32, 0, 0, 0, 1, 0, 0, 0 } example_pointer_deref={ 32, 0, 0, 0 } example_maybe_bad_cast_deref={ 32, 0, 0, 0, 0, 0, 0, 0 }

thread 941320 panic: attempt to use null value
/tmp/test.zig:39:39: 0x21ea26 in main (test)
    var val = example_maybe_bad_cast.*.?;
```

But this code is in a hot path and does work. It's because in the parent struct its an optional, so here when its cast back, the bits happen to align because in the current implementation the T is first, and it works:

```zig
const Parent = struct {
    example: ?Example,
};

var parent = Parent{ .example = example };
var parent_example_cast: *?Example = @ptrCast(&parent.example.?);

std.log.info("parent={any} parent_example_cast_deref={any}", .{
    std.mem.asBytes(&parent).*,
    std.mem.asBytes(parent_example_cast).*,
});

var val = parent_example_cast.*.?;
```

```
info: parent={ 32, 0, 0, 0, 1, 0, 0, 0 } parent_example_cast_deref={ 32, 0, 0, 0, 1, 0, 0, 0 }
```

Reply:
The "clean" solution is replacing the optional with an union, keeping the same functionality since Zig supports fieldParentPtr in union fields.

I think we should mark the code with a TODO(zig) because it's likely that optionals would be supported in fieldParentPtr some day.